### PR TITLE
Ec/auth backend2

### PR DESCRIPTION
**Pull request by exlxxn**
verified /user/signup and /user/login routes are working. sessionController.initializeSession also seems to be working but need to do more debugging. Have not yet verified sessionController.validateSession or sessionController.terminateSession.

### CODE REVIEW

File: server/controllers/userController.js [hunk 1]
```diff
@@ -1,25 +1,22 @@
...
-        WHERE NOT EXISTS (
-          SELECT email
-          FROM users
-          WHERE email = $2
-        )
+          (username, email, password, validateKey)
+        VALUES
+          ($1, $2, $3, $4)
+        ON CONFLICT (email) DO NOTHING
+        RETURNING *;
```
Comment (nattiechan) [inline server/controllers/userController.js]: Do you need _everything_ to be returned? Perhaps just `id`, or `username` would be sufficient

File: server/controllers/userController.js [hunk 2]
```diff
@@ -28,9 +25,12 @@ userController.createUser = (req, res, next) => {
         key
       ];
 
-      db.query(query, values);
+      return db.query(query, values);
     })
     .then(response => {
+      if (!response.rows.length) {
+        throw new Error('Username already taken');
```
Comment (nattiechan) [inline server/controllers/userController.js]: Should we throw an error or invoke the error handler instead?

File: server/controllers/userController.js [hunk 3]
```diff
@@ -44,24 +44,33 @@ userController.createUser = (req, res, next) => {
...
+    .then(response => {
+      const user = response.rows[0];
+      return bcrypt.compare(password, user.password);
+    })
+    .then(valid => {
+      if (valid) {
+        res.locals.user = username;
+        return next();
+      } else {
+        throw new Error('Invalid Credentials');
```
Comment (nattiechan) [inline server/controllers/userController.js]: same here

File: server/server.js [hunk 4]
```diff
@@ -12,6 +13,15 @@ const PORT = 3000;
 app.use(express.json());
 app.use(express.urlencoded({ extended: true }));
 
+/**
+ * session middleware
+ */
+app.use(session({
+  secret: 'mysuperduperubersecretsecret',
```
Comment (nattiechan) [inline server/server.js]: mayyyybeeeee we can put this as an `env` variable?

File: src/components/Signup.jsx [hunk 5]
```diff
@@ -20,25 +24,22 @@ function Signup() {
...
+      body: JSON.stringify(signupInfo)
+    })
+    .then(response => {
+      if (response.status === 200){
+        console.log('Sign up success!');
+        navigate('/');
+      }
+    })
+    .catch(err => {
+      console.log('Error: ', err);
```
Comment (nattiechan) [inline src/components/Signup.jsx]: In this case - should we `alert()` so users know what is happening instead?

File: server/server.js (hunk 4 above)
Comment (exlxxn) [reply server/server.js]: fixed. will be included in next commit.

File: server/controllers/userController.js (hunk 1 above)
Comment (exlxxn) [reply server/controllers/userController.js]: fixed. will be included in next commit.

File: server/controllers/userController.js (hunk 2 above)
Comment (exlxxn) [reply server/controllers/userController.js]: fixed. will be included in next commit.

File: server/controllers/userController.js (hunk 3 above)
Comment (exlxxn) [reply server/controllers/userController.js]: fixed. will be included in next commit.

File: src/components/Signup.jsx (hunk 5 above)
Comment (exlxxn) [reply src/components/Signup.jsx]: fixed. will be included in next commit.

File: src/components/Signup.jsx [hunk 6]
```diff
@@ -20,25 +24,25 @@ function Signup() {
...
+      body: JSON.stringify(signupInfo)
+    })
+    .then(response => {
+      if (response.status === 409) {
+        alert('Email already in use');
+      }
+      if (response.status === 200) {
+        alert('Signup Success!');
+        navigate('/');
+      }
```
Comment (nattiechan) [inline src/components/Signup.jsx]:
```suggestion
      if (response.status === 200) {
        alert('Signup Success!');
        navigate('/');
      } else if (response.status === 409) {
        alert('Email already in use');
      } else {
         alert('Signup failed. Please try again.')
      }
```

**Review by nattiechan — APPROVED** [review]

**Pull request merged**